### PR TITLE
== lowering needs to access private fields

### PR DESCRIPTION
**Pull request by WalterBright**
Blocking https://github.com/dlang/phobos/pull/6858

### CODE REVIEW

File: src/dmd/opover.d [hunk 1]
```diff
@@ -1269,7 +1269,11 @@ Expression op_overload(Expression e, Scope* sc)
                 if (!e.att2) e.att2 = t2;
                 e.e1 = new DotIdExp(e.loc, e.e1, Id._tupleof);
                 e.e2 = new DotIdExp(e.loc, e.e2, Id._tupleof);
-                result = e.expressionSemantic(sc);
+
+                auto sc2 = sc.push();
+                sc2.flags = (sc2.flags & ~SCOPE.onlysafeaccess) | SCOPE.noaccesscheck;
+                result = e.expressionSemantic(sc2);
+                sc2.pop();
```
Comment (jacob-carlborg) [inline src/dmd/opover.d]: Doesn't look like this is guarded by the `fieldwise` flag.

File: test/compilable/fieldwise.d [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+ */
+
+import imports.impfieldwise;
+
+@safe:
+
+bool test(S s, S t)
+{
+    return s == t; // comparison can access fields for ==
+}
```
Comment (jacob-carlborg) [inline test/compilable/fieldwise.d]: Do we have a test that verifies that it's not possible to compared the fields directory?

Comment (WalterBright) [reply test/compilable/fieldwise.d]: Yes, I found this problem because it was failing existing tests.

File: src/dmd/opover.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/opover.d]: That's right. The old behavior sometimes did fieldwise, and sometimes didn't. Apparently nobody reported it as a bug. Now it always does.

File: test/compilable/fieldwise.d [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/* REQUIRED_ARGS: -preview=dip1000 -preview=fieldwise
```
Comment (jacob-carlborg) [inline test/compilable/fieldwise.d]: If this change isn't guarded by `fieldwise`, do you still need the flag here?

**Comment by thewilsonator** [comment]
Is this good to go? CircleCI seems AWOL.

File: test/compilable/fieldwise.d (hunk 3 above)
Comment (WalterBright) [reply test/compilable/fieldwise.d]: Yes, because fieldwise changes the behavior before it.

**Comment by WalterBright** [comment]
It's good to go.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**